### PR TITLE
Disable automatic self updates in CI environments

### LINKED ISSUE #2886: error: rustup is not installed at '/home/user/.cargo' when $RUSTUP_HOME is set to something other than the former
**Problem**
1. Rustup cannot update when `CARGO_HOME` is not the Linux default `~/.cargo`
1. Then `rustup self uninstall` fails to work
1. At https://www.rust-lang.org/tools/install, there are no further instructions, or a link to those instructions, on how to uninstall `rustup`.

**Steps**

```bash
20:02 $ echo $RUSTUP_HOME
/home/<snip>/.rustup
✔ ~/src/<snip> [port ↑·2|✚ 2…1]
20:02 $ echo $CARGO_HOME
/home/<snip>/.local/share/cargo
^[[A✔ ~/src/<snip> [port ↑·2|✚ 2…1]
20:02 $ which rustup
/home/<snip>/.local/share/cargo/bin/rustup
✔ ~/src/<snip> [port ↑·2|✚ 2…1]
20:02 $ rustup self update
error: rustup is not installed at '/home/<snip>/.cargo'
```

**Possible Solution(s)**
No idea. We seem to be in a Mexican-standoff.

**Notes**

Output of `rustup --version`:

```bash
$ rustup --version
rustup 1.23.1 (3df2264a9 2020-11-30)
info: This is the version for the rustup toolchain manager, not the rustc compiler.
info: The currently active `rustc` version is `rustc 1.56.0-nightly (5ad7389bd 2021-08-06)`
```

Output of `rustup show`:

```bash
20:14 $ rustup show
Default host: x86_64-unknown-linux-gnu
rustup home:  /home/<snip>/.rustup

installed toolchains
--------------------

stable-x86_64-unknown-linux-gnu
nightly-x86_64-unknown-linux-gnu (default)

active toolchain
----------------

nightly-x86_64-unknown-linux-gnu (default)
rustc 1.56.0-nightly (5ad7389bd 2021-08-06)
```

### LINKED ISSUE #3824: Disable Rustup's self update by default if the `CI` environment variable is detected
Proposed as an alternative to #3821, mitigates #3709:

> Rustup could default to not auto update if the `CI` environment variable is present. This is simple both to implement and explain to users.

_Originally posted by @ChrisDenton in https://github.com/rust-lang/rustup/issues/3821#issuecomment-2106251789_

## Notes
- The relevant [user guide section](https://github.com/rust-lang/rustup/blob/454e8dc1bf31f20821cc8ac3228ea1f35b23d072/doc/user-guide/src/environment-variables.md) should be adjusted accordingly.
- This change might cause Rustup's own tests to fail, so removing `CI` from the environment variables is probably required before actually running those tests.

### PR DESCRIPTION
**Pull request by djc**
Fixes #2886, fixes #3824.

### CODE REVIEW

File: src/config.rs [hunk 1]
```diff
@@ -418,6 +418,10 @@ impl<'a> Cfg<'a> {
     }
 
     pub(crate) fn get_self_update_mode(&self) -> Result<SelfUpdateMode> {
+        if self.process.var("CI").is_ok() {
+            return Ok(SelfUpdateMode::Disable);
```
Comment (djc) [inline src/config.rs:424]: Alternatively, we could return `CheckOnly`?

Comment (rami3l) [reply src/config.rs:424]: @djc I guess disabled would still be better. It seems to me that the whole auto self update thing in the CI context is only introducing overheads.

Comment (djc) [reply src/config.rs:424]: That's what I thought as well. 👍

**Review by rami3l — APPROVED** [review]
Thanks! LGTM, as long as the CI is doing well...

File: src/config.rs [hunk 2]
```diff
@@ -418,6 +418,12 @@ impl<'a> Cfg<'a> {
     }
 
     pub(crate) fn get_self_update_mode(&self) -> Result<SelfUpdateMode> {
+        if self.process.var("CI").is_ok() {
+            if !self.process.var("RUSTUP_CI").is_ok() {
```
Comment (rami3l) [inline src/config.rs]: Can we flatten this nested if?

**Comment by smoelius** [comment]
FWIW, I think this is a really good idea. Thank you!

**Comment by hoodmane** [comment]
I see that this fixes the problem but rustup hasn't had a release in 9 months. Could rustup make a release with this commit?

**Comment by rami3l** [comment]
> I see that this fixes the problem but rustup hasn't had a release in 9 months. Could rustup make a release with this commit?

@hoodmane We are currently in the middle of releasing v1.28.0 (https://github.com/rust-lang/rustup/issues/3998, https://internals.rust-lang.org/t/seeking-beta-testers-for-rustup-v1-28-0). For the moment being we are actively collecting feedback on our beta release. Thank you for your patience :)

**Comment by hoodmane** [comment]
How do I install rustup beta?

**Comment by hoodmane** [comment]
I guess I should read your link and I'd find out.

**Comment by hoodmane** [comment]
Thanks @rami3l!

**Pull request merged**